pkg/ovalutil/rpm.go

```diff
@@ -0,0 +1,107 @@
+package ovalutil
+
+import (
+	"context"
+
+	"github.com/quay/claircore"
+	"github.com/quay/goval-parser/oval"
+	"github.com/rs/zerolog"
+)
+
+// ProtoVulnFunc allows a caller to create a prototype vulnerability that will be used
+// copied and further defined for every applicable oval.Criterion discovered.
+//
+// This allows the caller to use oval.Definition fields and closure syntax when
+// defining how a vulnerability should be parsed
+type ProtoVulnFunc func(def oval.Definition) (*claircore.Vulnerability, error)
+
+// RPMDefsToVulns iterates over the definitions in an oval root and assumes RPMInfo objects and states.
+//
+// Each Criterion encountered with an EVR string will be tranlated into a claircore.Vulnerability
+func RPMDefsToVulns(ctx context.Context, root oval.Root, protoVuln ProtoVulnFunc) ([]*claircore.Vulnerability, error) {
```

Comment:
I don't understand why the existing `RPMInfo.Extract` method wasn't extended to do this, as it already uses this sort of closure passing for creating vulnerability structs internally.

Reply:
This is simplified and easier to grok.

Reply:
Current method attempts to use OR and AND criteria prematurely, we need to work that into the entire application logic first. This is a refactor that grabs just what is necessary and keeps a slim DFS of the criterion. 

When we take a holistic look at Oval we can look back at the previous code which extracts AND and ORs.